src/main/java/io/logz/jmx2logzio/listener/prometheus/PromListenerWriter.kt

```diff
@@ -0,0 +1,123 @@
+import java.io.IOException
+import java.util.concurrent.Executors
+import java.util.concurrent.ScheduledExecutorService
+import java.util.concurrent.TimeUnit
+import javax.annotation.PostConstruct
+
+class PromListenerWriter(
+    senderParams: LogzioJavaSenderParams
+) : ListenerWriter {
+    companion object {
+        private val logger = LoggerFactory.getLogger(PromListenerWriter::class.java)
+    }
+
+    private val senderExecutors: ScheduledExecutorService = Executors.newScheduledThreadPool(senderParams.threadPoolSize);
+    private val sender: PromLogzioSender = createSender(senderExecutors, senderParams)!!.also {
+        it.start()
+    }
+
+    private fun createSender(senderExecutors: ScheduledExecutorService, logzioSenderParams: LogzioJavaSenderParams): PromLogzioSender? {
+        try {
+            val statusReporter = StatusReporterFactory.newSenderStatusReporter(LoggerFactory.getLogger(logzioSenderParams.loggerName))
+
+            val requestConf = HttpsRequestConfiguration
+                .builder()
+                .setLogzioListenerUrl(logzioSenderParams.url)
+                .setLogzioType(logzioSenderParams.type)
+                .setLogzioToken(logzioSenderParams.token)
+                .build()
+
+            val diskQueue = TimeSeriesDiskQueue(logzioSenderParams.queueDir,
+                reporter = statusReporter,
+                executorService = senderExecutors,
+                fsPercentThreshold = logzioSenderParams.fileSystemFullPercentThreshold,
+                gcPersistedQueueFilesIntervalSeconds = logzioSenderParams.gcPersistedQueueFilesIntervalSeconds)
+
+            return PromLogzioSender(statusReporter, senderExecutors, diskQueue, httpsRequestConfiguration = requestConf)
+        } catch (e: LogzioParameterErrorException) {
+            logger.error("problem in one or more parameters with error {}", e.message, e)
+        }
+
+        return null
+    }
+
+    override fun writeMetrics(metricDocs: List<Metric>) {
+        logger.debug("sending ${metricDocs.size} metrics")
+        try {
+            metricDocs.forEach {
+                val timeSeriesList = buildTimeSeries(it)
+                timeSeriesList.forEach { ts ->
+                    sender.send(ts.toByteArray())
+                }
+            }
+        } catch (e: IOException) {
+            e.printStackTrace()
+        }
+    }
+
+    @PostConstruct
+    override fun start() {
+        enableHangupSupport()
+    }
+
+    override fun shutdown() {
+        sender.stop()
+        senderExecutors.shutdown()
+        try {
+            senderExecutors.awaitTermination(20, TimeUnit.SECONDS)
+        } catch (e: InterruptedException) {
+            logger.error("Shutdown was interrupted {}", e.message, e)
+        }
+        if (!senderExecutors.isTerminated) {
+            senderExecutors.shutdownNow()
+        }
+        logger.info("Closing Listener Writer...")
+    }
+
+    private fun buildTimeSeries(metricDoc: Metric): List<TimeSeries> {
```

Comment:
it's not metricDoc. It's metric. The only problem is that it's a lie. It's not Metric single, but it contains multiple metrics, so it's in fact Metrics.

Reply:
that's why I called it metricDoc, since it's not a metric
all this code needs some refactoring...